content/blog/gargle-1-2-0/index.Rmd

```diff
@@ -0,0 +1,243 @@
+
+* Previously, httpuv was only indirectly suggested by gargle, via httr.
+  gargle now lists httpuv in its own `Suggests` field.
+* We encourage the installation of httpuv in interactive sessions, if we're
+  about to initiate oob auth, when it seems like we don't have to.
+
+## User interface
+
+The user interface has gotten more stylish, thanks to the cli package (<https://cli.r-lib.org>).
+This applies to informational messages, as well as to errors.
+
+All errors thrown by gargle now route through `rlang::abort()`, providing better access to the backtrace and error data.
+These errors have, at the very least, the `gargle_error` class and may also have additional subclasses.
+
+### Verbosity
+
+Since gargle is not a user-facing package, it has very little to say and only emits messages that end users *really* need to see.
+But sometimes it is useful to get more information about gargle's activities, especially when trying to figure out why `token_fetch()` isn't working the way you expect.
+In the past, we used the `"gargle_quiet"` option for this, which could be `TRUE` or `FALSE`.
+
+The previous on/off switch has been deprecated in favor of the new `"gargle_verbosity"` option, which has three levels:
+
+* `"debug"` is equivalent to the previous `gargle_quiet = FALSE`.
+  Use for debugging and troubleshooting.
+* `"info"` (the default) is basically equivalent to the previous
+  `gargle_quiet = TRUE`.
+* `"silent"` can be used to suppress all gargle messages. It has no previous
+  equivalent.
+
+There are a couple of helpers around verbosity:
+
+* `gargle_verbosity()` reports the current verbosity level.
+* `with_gargle_verbosity()` and `local_gargle_verbosity()` are
+  [withr-style](https://withr.r-lib.org) functions that temporarily modify the
+  verbosity level.
+
+## Retries
+
+gargle offers some Google-specific support for forming requests and handling responses, in addition to and separate from auth.
+
+Not all wrapper packages use this functionality, but, in particular, googlesheets4 *does*.
+The [usage limits for the Sheets API](https://developers.google.com/sheets/api/reference/limits) are low enough that regular folks run up against them fairly often.
+This aggravation is somewhat specific to Sheets, i.e. we don't see similar pain with the Drive and BigQuery APIs, which have much more generous limits.
+When you hit the Sheets "per user" or "per project" limit, requests fail with the error code `429 RESOURCE_EXHAUSTED`.[^3]
+
+[^3]: There is a limit of "500 requests per 100 seconds per project".
+In the case of a user token, the "project" refers to the GCP project that owns the OAuth client.
+Everyone who uses the built-in "Tidyverse API Packages" OAuth client is sharing this "per project" quota!
+If you're hitting "per project" limits regularly and it upsets your workflow, the universe is telling you that it's time to configure your own OAuth client or use a different method of auth.
```

Comment:
This seems important enough to go in the body?

Reply:
OK I'll move. I felt sheepish re: how Sheets-specific this was.